internal/controllers/image_controller.go

```diff
@@ -228,49 +228,191 @@ func (r *ImageReconciler) deleteImage(ctx context.Context, log logr.Logger, ioCt
+		}
+
+		if err := flattenChildImagesIfAny(log, ioCtx, img); err != nil {
+			return fmt.Errorf("failed to flatten snapshot child images: %w", err)
+		}
+
+		log.V(2).Info("Create snapshot clone", "snapshotId", snapName)
+		if err := r.cloneImageFromSnapshot(ctx, log, ioCtx, snapName, image); err != nil {
+			return fmt.Errorf("failed to create snapshot clone: %w", err)
+		}
+		log.V(2).Info("Created snapshot clone", "snapshotId", snapName)
+
+		if err := r.deleteImageSnapshot(ctx, log, ioCtx, snapName, snap); err != nil {
+			return fmt.Errorf("failed to delete image snapshot: %w", err)
+		}
 	}
-	log.V(2).Info("Volume image references", "pools", len(pools), "rbd-images", len(imgs))
 
-	if len(pools) != 0 && len(imgs) != 0 {
-		return fmt.Errorf("unable to delete volume image: still in use")
+	return nil
+}
+
+func (r *ImageReconciler) cloneImageFromSnapshot(ctx context.Context, log logr.Logger, ioCtx *rados.IOContext, clonedImageName string, image *providerapi.Image) error {
+	options := librbd.NewRbdImageOptions()
+	defer options.Destroy()
+	if err := options.SetString(librbd.ImageOptionDataPool, r.pool); err != nil {
+		return fmt.Errorf("failed to set data pool: %w", err)
 	}
 
-	if err := librbd.RemoveImage(ioCtx, ImageIDToRBDID(image.ID)); err != nil && !errors.Is(err, librbd.ErrNotFound) {
-		return fmt.Errorf("failed to remove rbd image: %w", err)
+	clonedImage := &providerapi.Image{
+		Metadata: apiutils.Metadata{
+			ID: clonedImageName,
+		},
+		Spec: providerapi.ImageSpec{
+			Size:        image.Spec.Size,
+			Limits:      image.Spec.Limits,
+			SnapshotRef: &clonedImageName,
+			Encryption:  image.Spec.Encryption,
+		},
 	}
-	log.V(2).Info("Rbd image deleted")
 
-	snapshotID := image.Spec.SnapshotRef
-	if image.Spec.Image != "" && snapshotID != nil && *snapshotID != "" {
-		log.V(2).Info("Deleting volume os-image")
-		if err := r.snapshots.Delete(ctx, *snapshotID); err != nil {
-			if !errors.Is(err, utils.ErrSnapshotNotFound) {
-				r.Eventf(image.Metadata, corev1.EventTypeWarning, "ImageDeletionFailed", "Error deleting image: %s", err)
-				return fmt.Errorf("error deleting os-image: %w", err)
+	log.V(2).Info("Creating image from snapshot", "snapshotId", clonedImageName)
+	_, err := r.createImageFromSnapshot(ctx, log, ioCtx, clonedImage, clonedImageName, options)
+	if err != nil {
+		return fmt.Errorf("failed to create image from snapshot: %w", err)
+	}
+
+	providerapi.SetManagerLabel(image, providerapi.VolumeManager)
+	if _, err := r.images.Create(ctx, clonedImage); err != nil {
+		return fmt.Errorf("failed to create image: %w", err)
+	}
+	return nil
+}
+
+func (r *ImageReconciler) deleteImageSnapshot(ctx context.Context, log logr.Logger, ioCtx *rados.IOContext, snapName string, snap *librbd.Snapshot) error {
+	// delete the cloned image if next steps fails
+	deleteClone := true
+	defer func() {
+		if deleteClone {
+			if err := r.images.Delete(ctx, snapName); err != nil {
```

Comment:
I don't think it's a good idea to have cleanup code here in this function for an image that gets created in a different function. If we ever refactor the code it might happen that we change the order we call functions and it might mess with this cleanup code.

Also, instead of deleting the cloned image in case of a failure, maybe we can keep track of the cloned image and reuse it when we retry the deletion.